jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/FileSystemPool.java

```diff
@@ -0,0 +1,273 @@
+//
+// ========================================================================
+// Copyright (c) 1995-2022 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.util.resource;
+
+import java.io.IOException;
+import java.net.URI;
+import java.nio.file.FileSystem;
+import java.nio.file.FileSystemAlreadyExistsException;
+import java.nio.file.FileSystemNotFoundException;
+import java.nio.file.FileSystems;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.attribute.FileTime;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.eclipse.jetty.util.IO;
+import org.eclipse.jetty.util.thread.AutoLock;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class FileSystemPool
+{
+    private static final Logger LOG = LoggerFactory.getLogger(FileSystemPool.class);
+    public static final FileSystemPool INSTANCE = new FileSystemPool();
+
+    private final Map<URI, Bucket> pool = new HashMap<>();
+    private final AutoLock poolLock = new AutoLock();
+
+    private FileSystemPool()
+    {
+    }
+
+    public Resource.Mount mount(URI uri) throws IOException
+    {
+        if (!uri.isAbsolute())
+            throw new IllegalArgumentException("not an absolute uri: " + uri);
+        if (PathResource.ALLOWED_SCHEMES.contains(uri.getScheme()))
+            throw new IllegalArgumentException("not an allowed scheme: " + uri);
+
+        FileSystem fileSystem = null;
+        try (AutoLock ignore = poolLock.lock())
+        {
+            try
+            {
+                fileSystem = FileSystems.newFileSystem(uri, Collections.emptyMap());
```

Comment:
Let's add the variant that takes a map when we need it as it's always easier to add some API than to remove it.

Where did you find about this `multi-release` property? It sounds like something useful to rewrite the `MultiReleaseJarFile` but that should be done in an extra PR.

Reply:
This is documented in the JDK.

For Java 9, it was the property key `multi-release`, and would use the JAR Implementation. `jdk.nio.zipfs.JarFileSystem`.

Starting in Java 17, it's now documented in  https://docs.oracle.com/en/java/javase/17/docs/api/jdk.zipfs/module-summary.html
The `JarFileSystem` was removed, in favor of rolling all of the logic into `ZipFsFileSystem`.
The old property name `multi-release` is still supported.
But the new property name `releaseVersion` was introduced to replace `multi-release`.
Same values are used in both cases (null/unset for no versioning, "runtime" for runtime versioning, how the default classloaders work, and a specific number for a specific version)

Reply:
Since this is for Jetty 12, which is Java 17 minimum, we should use ...

``` java
Map<String, String> env = new HashMap<>();
env.put("releaseVersion", "runtime");
```

Reply:
@joakime I'm guessing there is goodness in that PR, but let's get this one merged before @lorban goes on vacation and then your's should be rebased against HEAD and simpler to consider.